source/fuzz/transformation_move_instruction.cpp

```diff
@@ -0,0 +1,430 @@
+// Copyright (c) 2020 Vasyl Teliman
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_move_instruction.h"
+
+#include <queue>
+#include <unordered_set>
+#include <vector>
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationMoveInstruction::TransformationMoveInstruction(
+    const protobufs::TransformationMoveInstruction& message)
+    : message_(message) {}
+
+TransformationMoveInstruction::TransformationMoveInstruction(
+    const protobufs::InstructionDescriptor& insert_before,
+    const protobufs::InstructionDescriptor& target) {
+  *message_.mutable_insert_before() = insert_before;
+  *message_.mutable_target() = target;
+}
+
+bool TransformationMoveInstruction::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // |insert_before| must be valid.
+  auto* insert_before_inst =
+      FindInstruction(message_.insert_before(), ir_context);
+  if (!insert_before_inst) {
+    return false;
+  }
+
+  auto* target_inst = FindInstruction(message_.target(), ir_context);
+  if (!target_inst) {
+    return false;
+  }
+
+  return CanMoveInstruction(
+      ir_context,
+      fuzzerutil::GetIteratorForInstruction(
+          ir_context->get_instr_block(insert_before_inst), insert_before_inst),
+      fuzzerutil::GetIteratorForInstruction(
+          ir_context->get_instr_block(target_inst), target_inst));
+}
+
+void TransformationMoveInstruction::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto* insert_before_inst =
+      FindInstruction(message_.insert_before(), ir_context);
+  assert(insert_before_inst && "|insert_before| descriptor is invalid");
+
+  auto* target_inst = FindInstruction(message_.target(), ir_context);
+  assert(target_inst && "|target| descriptor is invalid");
+
+  assert(insert_before_inst != target_inst &&
+         "Can't insert target instruction before itself");
+
+  // Target instruction is not disposed after removal.
```

Comment:
Add ", thus it is OK to use it again below."

Reply:
Fixed in 587c227.